### PR TITLE
pkg/test: add new CreateWithFinalizer helper function

### PR DESCRIPTION
**Pull request by AlexNPavel**
This commit adds a new helper function that creates a provided
object and then creates a finalizer for the function to make sure
it is cleaned up properly. This is especially useful for single
namespace testing. The finalizer has a wait.PollImmediate to make
sure that the resource is fully deleted before returning

### CODE REVIEW

**Comment by AlexNPavel** [comment]
This is ready for review.

/cc @hasbro17 @fanminshi

File: pkg/test/context.go [hunk 1]
```diff
@@ -85,3 +100,40 @@ func (ctx *TestCtx) CleanupNoT() {
 func (ctx *TestCtx) AddFinalizerFn(fn finalizerFn) {
 	ctx.CleanUpFns = append(ctx.CleanUpFns, fn)
 }
+
+// TODO: figure out how to properly retrieve object kind info from object and enable
```
Comment (fanminshi) [inline pkg/test/context.go]: What's wrong with `obj.GetObjectKind()`?

File: pkg/test/context.go [hunk 2]
```diff
@@ -85,3 +100,40 @@ func (ctx *TestCtx) CleanupNoT() {
...
+	}
+	key, err := dynclient.ObjectKeyFromObject(obj)
+	//ctx.t.Logf("resource type %+v with namespace/name \"%+v\" created\n", obj.GetObjectKind(), key)
+	ctx.AddFinalizerFn(func() error {
+		err = Global.DynamicClient.Delete(gCtx, obj)
+		if err != nil {
+			return err
+		}
+		if cleanupOptions != nil && !cleanupOptions.SkipPolling {
+			return wait.PollImmediate(time.Second*1, time.Second*5, func() (bool, error) {
```
Comment (fanminshi) [inline pkg/test/context.go]: are we supposed to use the `cleanupOptions.Timeout ` and `cleanupOptions.finalizerRetryInterval` for `wait.PollImmediate(...)`?

File: pkg/test/context.go (hunk 1 above)
Comment (AlexNPavel) [reply pkg/test/context.go]: For some reason, on most object, the objectkind is blank for us. Not sure if why. Here is example output with the logging enabled:

```
Resource type &TypeMeta{Kind:,APIVersion:,} with namespace/name test-memcached/memcached-operator created
Resource type &TypeMeta{Kind:,APIVersion:,} with namespace/name test-memcached/memcached-operator created
Resource type &TypeMeta{Kind:,APIVersion:,} with namespace/name test-memcached/memcached-operator created
Resource type &TypeMeta{Kind:,APIVersion:,} with namespace/name test-memcached/memcached-operator created
Resource type &TypeMeta{Kind:,APIVersion:,} with namespace/name test-memcached/example-memcached created
```

Comment (AlexNPavel) [reply pkg/test/context.go]: The reason that we would need objectkind is because we have 4 different types of objects using the same name (service account, role, rolebinding, operator), so the output wouldn't the too helpful without the objectkind/typemeta

File: pkg/test/context.go (hunk 2 above)
Comment (AlexNPavel) [reply pkg/test/context.go]: Oops, yeah, forgot to change that when I added cleanupOptions

Comment (AlexNPavel) [reply pkg/test/context.go]: Fixed

File: pkg/test/context.go (hunk 1 above)
Comment (fanminshi) [reply pkg/test/context.go]: `For some reason, on most object, the objectkind is blank for us.` 
this makes sense because that usrs are not required to fill out the `TypeMeta` filed of the object they create and I believe `GetObjectKind()` simply retrieves that `TypeMeta. Kind ` field. 

I recall there is a way to infer the type of the runtime object because every interface has its actual type associate with it.

**Comment by fanminshi** [comment]
lgtm

File: pkg/test/context.go (hunk 1 above)
Comment (fanminshi) [reply pkg/test/context.go]: @AlexNPavel yeah, that's strange. I am unsure what's going on.

Comment (AlexNPavel) [reply pkg/test/context.go]: Hmm, just did a test and it seems that the controller runtime is modifying the object. It I do `GetObjectKind()` before the `client.Create()`, I get the correct `TypeMeta`. If I do it after the call to `client.Create()`, the `TypeMeta` is empty...

**Comment by fanminshi** [comment]
@AlexNPavel sure thing.

File: pkg/test/context.go [hunk 3]
```diff
@@ -85,3 +100,38 @@ func (ctx *TestCtx) CleanupNoT() {
...
+// finalizer function to delete it when Cleanup is called. In addition to the standard
+// controller-runtime client options
+func (ctx *TestCtx) CreateWithFinalizer(gCtx goctx.Context, obj runtime.Object, cleanupOptions *CleanupOptions) error {
+	objCopy := obj.DeepCopyObject()
+	err := Global.DynamicClient.Create(gCtx, obj)
+	if err != nil {
+		return err
+	}
+	key, err := dynclient.ObjectKeyFromObject(objCopy)
+	ctx.t.Logf("resource type %+v with namespace/name \"%+v\" created\n", objCopy.GetObjectKind().GroupVersionKind().Kind, key)
```
Comment (fanminshi) [inline pkg/test/context.go]: maybe replace `\"%+v\"` to `(%v)`?

File: pkg/test/context.go [hunk 4]
```diff
@@ -85,3 +100,38 @@ func (ctx *TestCtx) CleanupNoT() {
...
+		err = Global.DynamicClient.Delete(gCtx, objCopy)
+		if err != nil {
+			return err
+		}
+		if cleanupOptions != nil && !cleanupOptions.SkipPolling {
+			return wait.PollImmediate(cleanupOptions.RetryInterval, cleanupOptions.Timeout, func() (bool, error) {
+				err = Global.DynamicClient.Get(gCtx, key, objCopy)
+				if err != nil {
+					if apierrors.IsNotFound(err) {
+						ctx.t.Logf("resource type %+v with namespace/name \"%+v\" successfully deleted\n", objCopy.GetObjectKind().GroupVersionKind().Kind, key)
```
Comment (fanminshi) [inline pkg/test/context.go]: `"%+v"` to `(%v)`?

**Comment by fanminshi** [comment]
lgtm after nits.

File: pkg/test/context.go [hunk 5]
```diff
@@ -85,3 +100,38 @@ func (ctx *TestCtx) CleanupNoT() {
 func (ctx *TestCtx) AddFinalizerFn(fn finalizerFn) {
 	ctx.CleanUpFns = append(ctx.CleanUpFns, fn)
 }
+
+// CreateWithFinalizer uses the dynamic client to create an object and then adds a
+// finalizer function to delete it when Cleanup is called. In addition to the standard
+// controller-runtime client options
+func (ctx *TestCtx) CreateWithFinalizer(gCtx goctx.Context, obj runtime.Object, cleanupOptions *CleanupOptions) error {
```
Comment (hasbro17) [inline pkg/test/context.go]: After discussing it offline we're going to rename Finalizer to Cleanup in the context package. We're using a mix of both Finalizer and Cleanup already so we need to make the API consistent.
Plus Finalizers are already a k8s concept which might cause confusion on what the context cleanup is doing here.

So change this to `CreateWithCleanup`

**Comment by AlexNPavel** [comment]
Fixed

File: doc/test-framework/writing-e2e-tests.md [hunk 6]
```diff
@@ -101,19 +101,26 @@ The next step is to create a TestCtx for the current test and defer its cleanup
...
 
 ```go
-err := ctx.InitializeClusterResources()
+err := ctx.InitializeClusterResources(&framework.CleanupOptions{Timeout: time.Duration * 5, RetryInterval: time.Duration * 1})
 if err != nil {
     t.Fatalf("failed to initialize cluster resources: %v", err)
 }
 ```
 
+The `InitializeClusterResources` function uses the `CreateWithCleanup` function to creates the resources provided
```
Comment (hasbro17) [inline doc/test-framework/writing-e2e-tests.md]: `creates`==>`create`

File: doc/test-framework/writing-e2e-tests.md [hunk 7]
```diff
@@ -101,19 +101,26 @@ The next step is to create a TestCtx for the current test and defer its cleanup
...
+in your namespaced manifest. `CreateWithCleanup` use the controller-runtime's client to create resources and then
+creates a cleanup function that is called by `ctx.Cleanup` which deletes the resource and then waits for the resource to be
+fully deleted before returning. This is configurable with `CleanupOptions`. If the `CleanupOptions` argument is set
+to `nil`, the cleanup function simply calls the delete function and returns without waiting.
```
Comment (hasbro17) [inline doc/test-framework/writing-e2e-tests.md]: I know this is a bit subjective but how about rewording the above to:

`CreateWithCleanup` will create a resource and adds a cleanup function for that resource to the current context.
All cleanup functions for the context will run when `ctx.Cleanup()` called.
The cleanup function will delete the resource and then optionally poll and wait until the resource is completely removed from the apiserver. 
The `CleanupOptions` argument configures the polling timeout and retry period for the cleanup. 
If the `CleanupOptions` argument is set to `nil` the cleanup function will only delete and not wait for the resource to be removed from the apiserver.

File: test/e2e/incluster-test-code/memcached_test.go.tmpl [hunk 8]
```diff
@@ -100,8 +100,8 @@ func memcachedScaleTest(t *testing.T, f *framework.Framework, ctx *framework.Tes
 func MemcachedCluster(t *testing.T) {
 	t.Parallel()
 	ctx := framework.NewTestCtx(t)
-	defer ctx.Cleanup(t)
-	err := ctx.InitializeClusterResources()
+	defer ctx.Cleanup()
```
[the remark this answers is not in the document]
Comment (AlexNPavel) [reply test/e2e/incluster-test-code/memcached_test.go.tmpl]: We could do that. Currently, this PR allows a little bit more customization by allowing a user to choose different retry intervals and timeouts per object (except for namespaced resources, which all have the same values and global resources which we don't wait for), but it is a bit more verbose this way. Also, for cleanup we just run the cleanup functions in reverse order and create the cleanup functions immediately after the resource's creation. If we want to move the wait to cleanup, that may require some extra workarounds (cleanup is designed to do any cleanup, not just kube resource cleanup)

**Comment by AlexNPavel** [comment]
I tried to override the dynamic client earlier with a new interface but couldn't get it to work correctly, but now I figured it out, so making a `FrameworkClient` is definitely doable. For this to work correctly though, we would still need a new create function, because we can't override/disable the built in controller-runtime client's create with out own. I think it would make sense to decide on how to handle this now though, before merging this PR. Continually changing the syntax and forcing people to adapt every time we change something isn't ideal, even though this is still considered pre-alpha (a lot of people are already using this).

**Comment by hasbro17** [comment]
> For this to work correctly though, we would still need a new create function, because we can't override/disable the built in controller-runtime client's create with out own

@AlexNPavel We can either define a new FrameworkClient interface that has both `Create()` and `CreateWithCleanup()`
Or we just keep one `Create(ctx context.Context, object, cleanup options)`.
If we define and implement our own interface we don't have to override the dynamic client interface.

**Comment by AlexNPavel** [comment]
I've decided to close this PR and instead make a new PR that will create a `FrameworkClient`, encapsulating all the dynamic client's functions and putting the `CreateAndCleanup` functoinality in the framework client's `Create`.
